store/sqlstore/thread_store.go

```diff
@@ -52,247 +52,259 @@ func (s *SqlThreadStore) Get(id string) (*model.Thread, error) {
 	return &thread, nil
 }
 
-func (s *SqlThreadStore) GetThreadsForUser(userId, teamId string, opts model.GetUserThreadsOpts) (*model.Threads, error) {
-	type JoinedThread struct {
-		PostId         string
-		ReplyCount     int64
-		LastReplyAt    int64
-		LastViewedAt   int64
-		UnreadReplies  int64
-		UnreadMentions int64
-		Participants   model.StringArray
-		model.Post
+// GetTotalUnreadThreads counts the number of unread threads for the given user, optionally
+// constrained to the given team + DMs/GMs.
+func (s *SqlThreadStore) GetTotalUnreadThreads(userId, teamId string, opts model.GetUserThreadsOpts) (int64, error) {
+	var totalUnreadThreads int64
+
+	query := s.getQueryBuilder().
+		Select("COUNT(DISTINCT(Posts.RootId))").
+		From("Posts").
+		LeftJoin("ThreadMemberships ON Posts.RootId = ThreadMemberships.PostId").
+		Where("Posts.CreateAt > ThreadMemberships.LastViewed").
+		Where(sq.Eq{
+			"ThreadMemberships.UserId":    userId,
+			"ThreadMemberships.Following": true,
+		})
+
+	if teamId != "" {
+		query = query.
+			LeftJoin("Channels ON Posts.ChannelId = Channels.Id").
+			Where(sq.Or{
+				sq.Eq{"Channels.TeamId": teamId},
+				sq.Eq{"Channels.TeamId": ""},
+			})
 	}
 
-	fetchConditions := sq.And{
-		sq.Eq{"ThreadMemberships.UserId": userId},
-		sq.Eq{"ThreadMemberships.Following": true},
+	if !opts.Deleted {
+		query = query.Where(sq.Eq{"COALESCE(Posts.DeleteAt, 0)": 0})
 	}
 
+	sql, args, _ := query.ToSql()
```

Comment:
Let's check for error from these methods to be thorough :)

Reply:
Yes, good call -- I've pushed a change to address this.

Though I admit this used to be easier, when `GetMaster().Exec()` would accept the squirrel statement builder and do the `ToSQL` itself. Would we consider moving back to that model with the new `GetMasterX()`, and providing a back door as it were to run raw SQL?

Reply:
`GetMasterX` still has the same `Exec` method where you can pass raw sql. But AFAIR, `GetMaster().Exec()` didn't accept squirrel statement builder, only raw sqls. While converting the squirrel builder to raw sql, the error checking needs to happen _somewhere_. 0/5 if you want to move it somewhere else. But 4/5 on checking the error.

Reply:
Sounds good -- I'll likely follow up separately with a proposal that preserves the error handling but gets rid of the boilerplate.